### ISSUE TITLE
[gfx1100] bf16 ViT forward becomes non-deterministic and emits NaN above a sharp ~14.3k-token sequence-length threshold

### ISSUE THREAD

**Issue by AIwork4me** [ROCm/ROCm#6416]
## Summary

On **gfx1100 (RDNA3)** with the ROCm **torch 2.9.1** wheel (`hip 7.2.53211`) and **transformers 5.13.0**, the vision transformer (a standard pre-norm ViT using PyTorch SDPA attention) produces **non-deterministic output and `NaN`** when the per-image token/patch sequence length exceeds a **sharp threshold of ~14,200–14,688 tokens**. Below the threshold the same forward is **bit-for-bit deterministic** and the model produces correct output. The non-determinism is observed for a **deterministic-by-spec** forward (fixed weights, fixed input, `do_sample=False` / pure `argmax`).

This breaks end-to-end inference for high-resolution inputs: at full page resolution (~15k vision tokens) the model emits degenerate output (`"I"`, `"(1,0),(1000,999)"`, or a coherent-but-wrong paragraph on different identical runs) instead of correct OCR.

## Environment

- **GPU:** AMD Radeon gfx1100 (RDNA3), 48 GB (verified `rocminfo`: `gfx1100`)
- **ROCm:** userland `hip 7.2.53211-e1a6bc5663`; ROCk kernel module `6.14.14`
- **torch:** `2.9.1+gitff65f5b` (ROCm/HIP build), `torch.version.hip == 7.2.53211-e1a6bc5663`
- **transformers:** `5.13.0`
- **Model:** `tencent/HunyuanOCR` (HunyuanOCR-1.5, ~1 B params; native `hunyuan_vl` arch, standard pre-norm ViT vision tower with SDPA attention, `patch_size=16`, no QK-norm in the vision attention)
- OS: Linux x86_64; Python 3.12

## Reproducibility

Reproduces **every run** above the threshold; deterministic below it. The instability is in the **vision tower forward** (`model.model.get_image_features`), isolated as follows (all on the same GPU, bf16):

| Isolation experiment @ seq≈14,688 | Result |
|---|---|
| Full model ViT `get_image_features`, 3× identical | **non-deterministic** (`max|Δ|` up to 9312) and intermittently **NaN** |
| Bare `bf16` batched matmul `Q@Kᵀ` (heads×seq×hd) | **deterministic** (`max|Δ| = 0`), no NaN |
| Standalone `F.scaled_dot_product_attention` (bf16) | **deterministic**, no NaN |
| `nn.LayerNorm(1152)` on `[1, seq, 1152]` | **deterministic**, no NaN |
| Single real ViT block (`layers[9]`) with random input | **deterministic**, no NaN |
| LLM text-only forward @ 4,000 tokens | **deterministic**, no NaN |

So no single op reproduces it; the instability **emerges only in the full 27-layer ViT forward with the model's real weights/activations** at long sequence. The LLM is unaffected.

### Sequence-length threshold (full ViT `get_image_features`, 3× identical bf16 forwards)

| Image max-dim | ViT patches (seq len) | `max\|Δ\|` across 3 runs | NaN |
|---|---|---|---|
| ≤ 2272 | ≤ 14,200 | **0.000** | No (deterministic) |
| 2304 | 14,688 | **9312 / NaN** | **Yes** |
| full (2339) | 15,184 | **NaN** | **Yes** |

The transition is sharp: **deterministic at 14,200 tokens, non-deterministic + NaN at 14,688 tokens.** Just below the threshold (e.g. 13,056 tokens) end-to-end greedy decoding produces **correct OCR**; just above it produces garbage.

### End-to-end greedy (deterministic-by-spec) at full resolution

Three **identical** `do_sample=False` generations on the same full-resolution image yield three different outputs:

```
run0: 'I'
run1: '(1,0),(1000,999)'
run2: '# 2023年大学英语六级词汇选择与翻译\n\n## 词汇选择\n1. 首先'
```

## Minimal repro

```python
# Requires: pip install torch (ROCm 7.x wheel) "transformers==5.13.0" pillow
# Weights:  huggingface-cli download tencent/HunyuanOCR   (~2.2 GB)
import torch
from PIL import Image
from transformers import AutoProcessor, HunYuanVLForConditionalGeneration

MODEL = "tencent/HunyuanOCR"   # or local path
proc = AutoProcessor.from_pretrained(MODEL, use_fast=False)
model = HunYuanVLForConditionalGeneration.from_pretrained(
    MODEL, attn_implementation="sdpa", dtype=torch.bfloat16
).to("cuda:0").eval()

# any RGB image; thumbnail controls the ViT sequence length
img = Image.new("RGB", (3000, 3000))  # or a real page image

def vit_run(maxdim):
    im = img.copy(); im.thumbnail((maxdim, maxdim))
    inp = proc(images=im, return_tensors="pt").to("cuda:0")   # pixel_values + image_grid_thw
    feats = []
    with torch.inference_mode():
        for _ in range(3):
            f = model.model.get_image_features(
                inp["pixel_values"], inp["image_grid_thw"], return_dict=True
            )["last_hidden_state"].float().cpu()
            feats.append(f)
    d = max(torch.abs(feats[0]-feats[1]).max().item(),
            torch.abs(feats[0]-feats[2]).max().item())
    nan = any(torch.isnan(x).any().item() for x in feats)
    return inp["pixel_values"].shape[0], d, nan

for md in (2208, 2304):           # 2208 -> ~13.5k tokens (safe); 2304 -> ~14.7k tokens (triggers)
    n, d, nan = vit_run(md)
    print(f"maxdim={md} tokens={n} 3x-max-diff={d} NaN={nan}")
# Expected:
#   maxdim=2208 tokens=~13524 3x-max-diff=0.0   NaN=False
#   maxdim=2304 tokens=~14688 3x-max-diff=NaN(or large) NaN=True
```

## Notes / what I could not confirm

- **No NVIDIA GPU available** in this environment, so I could **not directly compare against CUDA** to prove ROCm-specificity. A run of the above on an NVIDIA GPU (where upstream reports the model works correctly at this resolution) would confirm whether this is a ROCm numerical issue vs a model-level instability.
- A single `fp32` forward at full resolution on gfx1100 does **not** NaN (it stays finite), which suggests the failure is specifically in the **bf16** long-sequence path. The **run-to-run non-determinism of a deterministic forward**, however, is a GPU/kernel behavior regardless of precision.
- The threshold (~14,300 tokens) does not correspond to an obvious power-of-two boundary; it may align with an internal tile/sequence limit in a HIP attention/GEMM kernel.

## Workaround

Cap the image resolution so the ViT sequence length stays below ~14,000 tokens (≈ `max_pixels ≈ 3.5e6`, i.e. longest side ≲ 1870 px). Below the threshold inference is deterministic and produces correct output.

## Ask

Could the ROCm team:
1. Reproduce the above on gfx1100 (RDNA3) and confirm whether the bf16 long-sequence ViT forward non-determinism + NaN is a HIP kernel issue;
2. If possible, compare against CUDA to confirm ROCm-specificity;
3. Advise whether a kernel/environment setting (e.g. forcing an fp32-accumulation attention path, or a newer ROCm/torch build) restores determinism and correctness at long sequence.

Happy to provide more isolation data or a narrower repro on request.

**Comment by schung-amd** [comment]
Thanks for the report. I think this is stale; I did reproduce this with the latest gfx110X-dgpu torch wheel in https://rocm.nightlies.amd.com/v2/gfx110X-dgpu/torch/, but that is very outdated at the moment, and this issue no longer reproduces on my end with the latest gfx110X-all wheel in https://rocm.nightlies.amd.com/v2/gfx110X-all/torch/. Do you need this working on the old wheel specifically?

**Comment by AIwork4me** [comment]
@schung-amd Thank you for investigating this and pointing us to the `gfx110X-all` wheel.

We've verified end-to-end: the bug reproduces on the old `gfx110X-dgpu` wheel and is completely fixed on the latest `gfx110X-all` wheel.

### Verification results

| ViT grid | Tokens | `gfx110X-dgpu` (torch 2.9.1, hip 7.2) | `gfx110X-all` (torch 2.10.0, hip 7.12) |
|----------|--------|----------------------------------------|----------------------------------------|
| 112x128  | 14,336 | deterministic (&vert;&Delta;&vert;=0.000) | deterministic (&vert;&Delta;&vert;=0.000) |
| 116x128  | 14,848 | **non-deterministic (&vert;&Delta;&vert;=11,315)** | deterministic (&vert;&Delta;&vert;=0.000) |
| 120x128  | 15,360 | **non-deterministic (&vert;&Delta;&vert;=11,112)** | deterministic (&vert;&Delta;&vert;=0.000) |
| 124x128  | 15,872 | **non-deterministic (&vert;&Delta;&vert;=11,053)** | deterministic (&vert;&Delta;&vert;=0.000) |
| 128x128  | 16,384 | deterministic (&vert;&Delta;&vert;=0.000) | deterministic (&vert;&Delta;&vert;=0.000) |

*Test methodology: 10 identical bf16 ViT forward passes per grid, HunYuanOCR model (tencent/HunyuanOCR-1.5), `F.scaled_dot_product_attention` attention, `max|Δ|` computed pairwise across all 10 runs. All tests on single gfx1100 GPU, `torch.inference_mode()`.*

### Environment

| | `gfx110X-dgpu` (broken) | `gfx110X-all` (fixed) |
|---|---|---|
| torch | `2.9.1+gitff65f5b` | `2.10.0+rocm7.12.0a20260311` |
| HIP version | `7.2.53211-e1a6bc5663` | `7.12.60610` |
| ROCk kernel | `6.14.14` | `6.14.14` (unchanged) |
| GPU | gfx1100 (RDNA3), 48 GB | same |
| transformers | `5.13.0` | `5.13.0` |
| Wheel source | `rocm.nightlies.amd.com/v2/gfx110X-dgpu/torch/` | `rocm.nightlies.amd.com/v2/gfx110X-all/torch/` |

### Notes

1. **No NaN observed** on the old wheel in our reproduction — pure non-determinism (max|Δ| ~11k), with 99.9% of output elements differing between runs. The NaN in the original report may be a downstream accumulation effect from the non-deterministic ViT output.
2. **The 128x128 case was deterministic on both wheels.** This suggests the bug is in a specific tile-size or workload-size kernel dispatch path, not a uniform degradation at large token counts.
3. **To answer your question: we do not need backporting to the old `gfx110X-dgpu` wheel.** We are migrating to `gfx110X-all`. You are welcome to close this issue.
4. **One question:** Is there a **minimum ROCm kernel driver version required** for the `gfx110X-all` wheel? Ours works at `6.14.14`, but documenting the minimum for downstream users would help.

### Minimal repro (portable across wheels)

<details>
<summary>repro.py</summary>

```python
# Install torch from gfx110X-dgpu or gfx110X-all, then:
# pip install "transformers==5.13.0" pillow
import torch
from PIL import Image

MODEL_PATH = "tencent/HunyuanOCR"

from transformers import AutoProcessor, HunYuanVLForConditionalGeneration
proc = AutoProcessor.from_pretrained(MODEL_PATH, use_fast=False)
model = HunYuanVLForConditionalGeneration.from_pretrained(
    MODEL_PATH, attn_implementation="sdpa", torch_dtype=torch.bfloat16
).to("cuda:0").eval()

for w, h in [(2048, 1792), (2048, 1856), (2048, 1920), (2048, 1984), (2048, 2048)]:
    img = Image.new("RGB", (w, h))
    with torch.inference_mode():
        inp = proc(images=img, return_tensors="pt").to("cuda:0")
        thw = inp["image_grid_thw"]
        tokens = int(thw[0,1] * thw[0,2])
        feats = []
        for _ in range(10):
            f = model.model.get_image_features(
                inp["pixel_values"], inp["image_grid_thw"], return_dict=True
            )["last_hidden_state"].float().cpu()
            feats.append(f)
        max_d = max(
            torch.abs(feats[i] - feats[j]).max().item()
            for i in range(10) for j in range(i + 1, 10)
        )
        print(f"tokens={tokens:>5d} max|Δ|={max_d:>10.4f} deterministic={max_d==0}")
```
</details>

Thanks again for the quick response and pointing us to the updated wheel channel.

**Comment by AIwork4me** [comment]
@schung-amd Quick follow-up:

1. We've published an installer that wraps the `gfx110X-all` stack for AMD RDNA3 GPUs, making the fix from this issue trivially reproducible:
   **https://github.com/AIwork4me/rocm-vllm-installer**

   One command installs torch 2.10 (gfx110X-all) + vLLM 0.19.0 (from source) + all patches for torchvision/ROCm compatibility. Verified end-to-end: vLLM serving + HunyuanOCR, deterministic at 16.4k tokens.

2. **Question on stable releases:** Is there a GA/stable release of ROCm 7.12 (or the `gfx110X-all` torch wheel) available on `download.pytorch.org`? Currently we pin to `rocm.nightlies.amd.com/v2/gfx110X-all/` which works, but we'd like to point users at a stable artifact once one exists. Is there a public timeline for when the fix will land in an official release?

Thanks again for the quick confirmation of the fix.

**Comment by schung-amd** [comment]
>One question: Is there a minimum ROCm kernel driver version required for the gfx110X-all wheel? Ours works at 6.14.14, but documenting the minimum for downstream users would help.

I don't think we've established a minimum kernel version per se, but TheRock wheels should support anything from the version outlined in https://rocm.docs.amd.com/en/latest/compatibility/compatibility-matrix.html and onward.

>Question on stable releases: Is there a GA/stable release of ROCm 7.12 (or the gfx110X-all torch wheel) available on download.pytorch.org? Currently we pin to rocm.nightlies.amd.com/v2/gfx110X-all/ which works, but we'd like to point users at a stable artifact once one exists. Is there a public timeline for when the fix will land in an official release?

I haven't personally tried the repro on this wheel yet, but https://rocm.docs.amd.com/projects/ai-ecosystem/en/latest/frameworks/pytorch/install.html has a pip install for torch + ROCm 7.14. There are also other 7.14 and 7.13 wheels at https://repo.amd.com/rocm/whl-multi-arch/ if one suits your needs better.

Closing for now, but feel free to comment if these wheels don't work or if you need further guidance and we can reopen if necessary!

**Comment by AIwork4me** [comment]
@schung-amd Confirmed on the **stable ROCm 7.14 GA** wheel you linked (the one you hadn't tested yet): it fixes #6416 end-to-end on gfx1100.

**Env:** gfx1100 / RDNA3 (Radeon Pro W7900D), ROCk `6.14.14`, `transformers==5.13.0`, HunyuanOCR-1.5, `attn_implementation="sdpa"`, bf16.

**Install (lean — `device-gfx1100` is enough, no `device-all` needed):**

```bash
pip install --index-url https://repo.amd.com/rocm/whl-multi-arch/ \
  "torch[device-gfx1100]==2.11.0+rocm7.14.0" \
  "torchvision[device-gfx1100]==0.26.0+rocm7.14.0" \
  "torchaudio==2.11.0+rocm7.14.0"
```
→ `torch 2.11.0+rocm7.14.0`, `hip 7.14.60850`, `torch.cuda.is_available()=True` on the 6.14.14 kernel.

**ViT forward, 10× identical bf16 passes (pairwise `max|Δ|`):**

| ViT grid | tokens | torch 2.9.1+rocm7.2 (hip 7.2) | **7.14 GA (hip 7.14)** |
|---|---|---|---|
| 116×128 | 14,848 | 576 | **0** |
| 120×128 | 15,360 | 8.7e30 | **0** |
| 124×128 | 15,872 | 576 | **0** |
| 112×128 / 128×128 | 14,336 / 16,384 | 0 / 0 | 0 / 0 |

**End-to-end `generate(do_sample=False)` ×3 @ 15,360 tokens, same image:**
- **2.9.1+rocm7.2 →** 3 divergent/degenerate outputs: `图片中没有文字。` / `I I I I…` / `图片中的文本内容是：I I I I…`
- **7.14 GA →** 3× byte-identical, OCR verbatim-correct.

**For others on RDNA3:** the bug is gone on the **GA** 7.14 channel — no need to pin the nightly `gfx110X-all` build. The old torch 2.9.1 / hip 7.2 wheel still carries it; migrate. The <14k-token workaround cap is no longer needed once on 7.14. Thanks again for the pointer.
